### PR TITLE
payout debt management

### PR DESCRIPTION
**Pull request by aie0**
- management fees
- node rewards

### CODE REVIEW

File: pallets/ddc-payouts/src/lib.rs [hunk 1]
```diff
@@ -235,8 +254,9 @@ pub mod pallet {
 		NotInitialized,
```
Comment (yahortsaryk) [inline pallets/ddc-payouts/src/lib.rs:254]: As we changed `ValueQuery` to `OptionQuery`, maybe we should remove the `NotInitialized` status now ?

**Review by yahortsaryk — DISMISSED** [review]

File: pallets/ddc-payouts/src/lib.rs [hunk 2]
```diff
@@ -187,30 +213,23 @@ pub mod pallet {
 
 	#[pallet::storage]
 	#[pallet::getter(fn debtor_customers)]
-	pub type DebtorCustomers<T: Config> = StorageDoubleMap<
```
Comment (Raid5594) [inline pallets/ddc-payouts/src/lib.rs:190]: I can see that we're adding debt to this structure, but how is it updated afterwards? For example the owner top ups the account and continues using the service

Comment (Raid5594) [reply pallets/ddc-payouts/src/lib.rs:190]: also, please rename customer_dept to customer_debt, it's a typo probably

File: pallets/ddc-payouts/src/mock.rs [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+}
+
+pub struct TestCustomerCharger;
+impl<T: Config> CustomerCharger<T> for TestCustomerCharger {
+	fn charge_content_owner(
+		_content_owner: T::AccountId,
+		_billing_vault: T::AccountId,
+		amount: u128,
+	) -> sp_runtime::DispatchResult {
+		ensure!(amount > 1_000_000_000, BadOrigin); //  any error will do
```
[the remark this answers is not in the document]
Comment (aie0) [reply pallets/ddc-payouts/src/mock.rs:117]: this is test so for some scenarios I want the function to fail to see how it reacts to it, so I set condition for failing the function

File: pallets/ddc-payouts/src/lib.rs [hunk 4]
```diff
@@ -133,6 +142,21 @@ pub mod pallet {
 			cluster_id: ClusterId,
 			era: DdcEra,
 		},
+		TreasuryFeesCharged {
```
[the remark this answers is not in the document]
Comment (aie0) [reply pallets/ddc-payouts/src/lib.rs:145]: good idea, but let's merge it as we already have 2 parallel branches to deal with changes

File: pallets/ddc-payouts/src/lib.rs (hunk 1 above)
Comment (aie0) [reply pallets/ddc-payouts/src/lib.rs:254]: we can merge several status together as part of optimisations - we'll deal with it as well once the time comes

File: pallets/ddc-payouts/src/lib.rs (hunk 2 above)
Comment (aie0) [reply pallets/ddc-payouts/src/lib.rs:190]: debt is not managed and definately not here, as it should be delt in the customers pallet when the owner tops up

File: pallets/ddc-payouts/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (aie0) [reply pallets/ddc-payouts/src/lib.rs:145]: making changes anyway - so might as well change it

Done

File: pallets/ddc-payouts/src/lib.rs (hunk 1 above)
Comment (yahortsaryk) [reply pallets/ddc-payouts/src/lib.rs:254]: ok, let's leave it like this for now

**Review by yahortsaryk — APPROVED** [review]

File: pallets/ddc-payouts/src/lib.rs (hunk 2 above)
Comment (Raid5594) [reply pallets/ddc-payouts/src/lib.rs:190]: the structure is a mapping between with keys: clusterid + accountid. When topping up the account, it seems to be inefficient to check all existing clusters. 

Maybe it's better to update ledger structure in account directly? 

The solution I envisioned initially, is to depleted the customer balance until it's 0, and add to the negative balance all of the debt. 
Then when we top up, we would cover the debt first, and only then add to the positive balance

Comment (aie0) [reply pallets/ddc-payouts/src/lib.rs:190]: currently we're only logging it and later think about what to actually do with it - for now there is no clear decision about it. the idea about negative balance is welcome and will be discussed as an alternative

**Review by Raid5594 — APPROVED** [review]

**Pull request merged**